### ISSUE TITLE
[otbn] Multi-bit encoding for CMD register?

### ISSUE THREAD

**Issue by imphil** [lowRISC/opentitan#7822]
OTBN has a command register (CMD). Discuss if this register should be multi-bit encoded.

**Comment by cdgori** [comment]
Referring to the published docs, CMD reg has only 1 bit - START?  Assuming that is still correct:

-START is presumably ignored by OTBN once an operation is in process.  No particularly interesting threat glitching it then.  (Please confirm it's ignored though.)

-START being glitched before all of the cryptographic data is transferred/available to OTBN could present some issues (incomplete RSA/ECC keys, signing of truncated payloads, or possibly executing truncated code on OTBN that would manipulate sensitive values in unsafe ways).  I would have to think for a long time about how realistic these scenarios are, it's probably better to just make START a nibble (write "A" to start) and be done with it.  

This effectively pushes the glitch-sensitive spot to a decoded version of this start condition though, so the value is marginal, and we would be better off somehow conditioning acceptance of START on (keys_valid && data_payload_completely_loaded && code_completely_loaded), at which point I probably wouldn't care if START was single or multi-bit.

Happy to hear other opinions here ( @moidx  ) or if there's a threat I'm missing, which is quite possible.

**Comment by moidx** [comment]
Hi @cdgori this is issue is also in reference to lowrisc/opentitan#7628. To my knowledge, we don't have a way to quantify `data_payload_completely_loaded` and `code_completely_loaded`, this is up to Ibex and the use of the control interface. I could be missing some details but @imphil  can chime in.

**Comment by imphil** [comment]
Thanks a lot @cdgori. I'm sorry I wasn't more clear, this issue was a spin-off from https://github.com/lowRISC/opentitan/pull/7628, where CMD becomes a register with three possible command values (execute, and 2x secure wipe). 

> -START is presumably ignored by OTBN once an operation is in process. No particularly interesting threat glitching it then. (Please confirm it's ignored though.)

Yes, it is ignored, and it's explicitly marked as such in the updated docs.

> -START being glitched before all of the cryptographic data is transferred/available to OTBN 

That's a good scenario. But as you say, if we make CMD multi-bit, the attack point just moves down into the place where the signal becomes single-bit. Wouldn't it be more reliable for software to just use a slightly different calling convention when calling OTBN to something like this:

  1. Ibex writes all crypto constants to OTBN DMEM.
  2. Ibex writes a single word to OTBN DMEM indicating "everything has been written"
  3. Ibex starts OTBN through CMD
  4. OTBN software checks if the "everything has been written" DMEM word is present.
  5. Now the normal crypto routine is executed.

This way the software author can determine the level of protection for the passed input parameters. They can even calculate a CRC32 or something like that over the input data and have OTBN software recalculate and check it.

**Comment by imphil** [comment]
> To my knowledge, we don't have a way to quantify data_payload_completely_loaded and code_completely_loaded, this is up to Ibex and the use of the control interface. I could be missing some details but @imphil can chime in.

That is correct from a hardware point of view -- there is no required calling convention for OTBN, and hence OTBN doesn't know when "all" input parameters have been written to DMEM. Software knows, however, and that's why such a check can be done in software, as indicated in the [previous comment](https://github.com/lowRISC/opentitan/issues/7822#issuecomment-901485556).
